R/data.table.R

```diff
@@ -1385,7 +1385,8 @@ replace_dot_alias = function(e) {
     byval = i
     bynames = if (missing(on)) head(key(x),length(leftcols)) else names(on)
     allbyvars = NULL
-    bysameorder = haskey(i) || (is.sorted(f__) && ((roll == FALSE) || length(f__) == 1L)) # Fix for #1010
+    bysameorder = (haskey(i) && identical(leftcols, head(match(key(i),names(i)), length(leftcols)))) || # leftcols leading subset of key(i); see #4917
```

Comment:
is there a reason for identical() over all( . == . ) here? identical is a heavier test including things like attributes right?

also at a glance I would do head() before match()

Reply:
Good point. `==` is an alloc returning logical vector which `all()` then passes over. It's the alloc that's nice to avoid. Usually fast but could trigger a gc.  Identical is a binary compare so, when there aren't any attributes to check, it's faster than all.equal.  I seem to remember some tests from years back posted somewhere and `identical` was fastest.  But could be wrong and would be surprised if there's much in it.
Yep head before match good idea. Will do.

Reply:
caveat ... `==` for a length 1 vector is a special case that doesn't allocate, just for the reason that R has global internal constants for TRUE, FALSE, and NA_LOGICAL.  So the result of  `x==1L` when x is length 1L will be one of those global pointers without an alloc.  As always, IIUC.